File: templates/javascript/v1/state/states.js
export const draftState = {
  name: "draft",
  handle(input) {
    return `DRAFT: ${input}`;
  }
};

export const publishedState = {
  name: "published",
  handle(input) {
    return `PUBLISHED: ${input.toUpperCase()}`;
  }
};

export class DocumentContext {
  constructor(state) {
    this.state = state;
  }

  setState(state) {
    this.state = state;
  }

  process(input) {
    return this.state.handle(input);
  }

  currentState() {
    return this.state.name;
  }
}

File: templates/python/v1/state/states.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol


class State(Protocol):
    def handle(self, input_text: str) -> str:
        ...

    def name(self) -> str:
        ...


@dataclass(slots=True)
class DraftState:
    def handle(self, input_text: str) -> str:
        return f"DRAFT: {input_text}"

    def name(self) -> str:
        return "draft"


@dataclass(slots=True)
class PublishedState:
    def handle(self, input_text: str) -> str:
        return f"PUBLISHED: {input_text.upper()}"

    def name(self) -> str:
        return "published"


@dataclass(slots=True)
class DocumentContext:
    state: State

    def set_state(self, state: State) -> None:
        self.state = state

    def process(self, input_text: str) -> str:
        return self.state.handle(input_text)

    def current_state(self) -> str:
        return self.state.name()
